### PR TITLE
Replacing session-based authentication with JWTs.

### PR DESCRIPTION
**Pull request by 0xf8f8ff**
This PR replaces HTTP sessions with JSON Web Tokens for user authentication.

Note: this PR implements some basic functionality for refreshing JWTs, but the main feature will be added in the following PR dedicated to the new (interceptor-based) access control process.

### CODE REVIEW

File: web/auth/interceptors/request_validator.go
Comment (meling) [inline web/auth/interceptors/request_validator.go:31]: Not sure we discussed it before, but the prometheus metrics stuff seems like a good target for a separate interceptor in `metrics.go`. Also, the `qf.Ag*` types should probably be moved to the interceptor package instead of the `qf` package? 

Similarly, I wonder if the following:
```go
	reg.MustRegister(
		promgrpc.NewServerMetrics(),
		qf.AgFailedMethodsMetric,
		qf.AgMethodSuccessRateMetric,
		qf.AgResponseTimeByMethodsMetric,
	)
```
could be move to the interceptor package's `init()` func. You would need to add a func like this:
```go
var reg = prometheus.NewRegistry()

func NewMetricsRegistry() *prometheus.Registry {
  return reg
}
```
So that you can get the `reg` to be passed to the http server. Or you could move this too:
```go
// Create an HTTP server for prometheus.
	httpServer := &http.Server{
		Handler: promhttp.HandlerFor(reg, promhttp.HandlerOpts{}),
		Addr:    fmt.Sprintf("127.0.0.1:%d", 9097),
	}
	go func() {
		if err := httpServer.ListenAndServe(); err != nil {
			log.Fatalf("Failed to start a http server: %v", err)
		}
	}()
```
And provide some other way to start it from main... but be aware of the `log.Fatalf` not being in `main.go` is an issue...

Comment (JosteinLindhom) [reply web/auth/interceptors/request_validator.go:31]: I split the interceptors in this branch [grpc-interceptors](https://github.com/quickfeed/quickfeed/tree/grpc-interceptors).

Metrics: [metrics.go](https://github.com/quickfeed/quickfeed/blob/grpc-interceptors/qf/metrics.go)
Validation: [validation.go](https://github.com/quickfeed/quickfeed/blob/grpc-interceptors/qf/validation.go)
User Verifier: [auth.go](https://github.com/quickfeed/quickfeed/blob/83054cfbacd03b341db6abc85098f9703db507b4/web/auth/auth.go#L510)

Usage in `main.go`: [link](https://github.com/quickfeed/quickfeed/blob/83054cfbacd03b341db6abc85098f9703db507b4/main.go#L107)

`auth.go` contains both unary and streaming variants of the interceptor.

Comment (0xf8f8ff) [reply web/auth/interceptors/request_validator.go:31]: Ok, should I then wait until it is ready and merged into master (and work on something else meanwhile)?

File: web/auth/auth_test.go
Comment (meling) [inline web/auth/auth_test.go:47]: We should ideally use a global/exported `const` instead of `"auth"`. There are two:
- `auth.CookieName`
- `tokens.authCookieName`

Suggest to rename the latter and use it as `tokens.CookieName` or `tokens.AuthCookieName`. This is because `auth` cannot be imported into `tokens` package.

File: web/auth/tokens/jwt.go
Comment (meling) [inline web/auth/tokens/jwt.go:1]: We should avoid packages with plural names. So maybe just `token` works? Or maybe it could be `jwt` (I prefer `token`). Another options is to move the content of the `tokens` package into the `auth` package.

File: web/auth/tokens/jwt_test.go
Comment (meling) [inline web/auth/tokens/jwt_test.go:13]: General comment about tests. I didn't look at these tests carefully in terms of the following. But I think it is possible that there are some benefits to be had from rephrasing the tests as table-driven tests. Things like:

```go
tests := []struct{
  inputA string
  inputB string
  want string
}{ ...
```

File: web/auth/tokens/update.go
Comment (meling) [inline web/auth/tokens/update.go:19]: Remember to add `.` at the end of doc comments.

File: web/interceptor/helpers.go
Comment (meling) [inline web/interceptor/helpers.go:39]: Should the `Set-Cookie` use a const?

File: web/interceptor/helpers.go
Comment (meling) [inline web/interceptor/helpers.go:26]: The returned error from `extractToken` is already `ErrInvalidAuthCookie` and can be returned directly.

File: web/interceptor/helpers.go
Comment (meling) [inline web/interceptor/helpers.go:37]: Should we continue after this fail? or should the error be returned?

File: web/interceptor/auth_test.go
Comment (meling) [inline web/interceptor/auth_test.go:111]: `s/GetSelf/GetUser/`

**Review by meling — CHANGES_REQUESTED** [review]
Mostly minor... Main issue is package rename or reorganizing.

File: web/auth/tokens/jwt.go
Comment (0xf8f8ff) [reply web/auth/tokens/jwt.go]: Moved into the `auth` package.

File: web/auth/tokens/jwt_test.go
Comment (0xf8f8ff) [reply web/auth/tokens/jwt_test.go]: Does it really apply to the tests in this file? I agree in general but can't see how it could be beneficial in this case.

File: web/interceptor/helpers.go
Comment (0xf8f8ff) [reply web/interceptor/helpers.go]: It absolutely should. Fixed.

File: web/interceptor/helpers.go
Comment (0xf8f8ff) [reply web/interceptor/helpers.go:36]: I think it is better to return an error. Fixed.

**Pull request merged**
